hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/MappingJdbcSqlConnectorTest.java

```diff
@@ -305,6 +306,59 @@ public void when_mappingIsDeclaredWithDataConnection_then_itIsAvailable() throws
         assertThat(object).isEqualTo(name);
     }
 
+    /**
+     * Source : <a href=https://github.com/hazelcast/hazelcast/issues/24337">issue #24337</a>.
+     */
+    @Test
+    public void given_mappingIsDeclaredWithDataConn_when_DataConnWasRemoved_then_success() throws Exception {
+        // given
+        String dcName = randomName();
+        String mappingName = randomName();
+        createTable(mappingName);
+        Map<String, String> options = new HashMap<>();
+        options.put("jdbcUrl", dbConnectionUrl);
+
+        // when
+        createDataConnection(instance(), dcName, "JDBC", false, options);
+        createJdbcMappingUsingDataConnection(mappingName, dcName);
+        sqlService.execute("DROP DATA CONNECTION " + dcName);
+
+        // then
+        assertRowsAnyOrder("SHOW DATA CONNECTIONS ", singletonList(new Row(TEST_DATABASE_REF)));
+
+        // Ensure that engine is not broken after Data Connection removal with some unrelated query.
+        assertRowsAnyOrder("SHOW MAPPINGS ", singletonList(new Row(mappingName)));
+
+        // Mapping shouldn't provide data, since data connection was removed.
+        assertThatThrownBy(() -> sqlService.execute("SELECT * FROM " + mappingName))
+                .hasMessageContaining("com.hazelcast.core.HazelcastException: Data connection '"
+                        + dcName + "' not found");
+    }
+
+    @Test
+    public void when_dataConnectionIsAltered_then_mappingsConnectorTypesAreUpdated() throws Exception {
```

Comment:
How about a query? Is a cached query correctly re-planned?

Reply:
good catch: no, plan cache is not invalidated. I will possibly need to add `DataConnectionListener`, similar to `TableListener`

Reply:
implemented, worth re-review